src/compiler/parser.ts

```diff
@@ -3617,11 +3715,31 @@ module ts {
             // FunctionExpression:
             //      function BindingIdentifieropt(FormalParameters) { FunctionBody }
             var node = <FunctionExpression>createNode(SyntaxKind.FunctionExpression);
+            setModifiers(node, parseModifiers());
             parseExpected(SyntaxKind.FunctionKeyword);
+            var isAsync = !!(node.flags & NodeFlags.Async);
             node.asteriskToken = parseOptionalToken(SyntaxKind.AsteriskToken);
-            node.name = node.asteriskToken ? doInYieldContext(parseOptionalIdentifier) : parseOptionalIdentifier();
-            fillSignature(SyntaxKind.ColonToken, /*yieldAndGeneratorParameterContext:*/ !!node.asteriskToken, /*requireCompleteParameterList:*/ false, node);
-            node.body = parseFunctionBlock(/*allowYield:*/ !!node.asteriskToken, /* ignoreMissingOpenBrace */ false);
+
+            if (node.asteriskToken) {
+                var savedYieldContext = inYieldContext();
+                setYieldContext(true);
+            }
+            if (isAsync) {
+                var savedAwaitContext = inAwaitContext();
+                setAwaitContext(true);
+            }
+
+            node.name = parseOptionalIdentifier();

```

Comment:
can this be made simpler with your new functions to set context, parse, and then restore context?